modules/library/main/src/main/java/org/geotools/data/DataUtilities.java

```diff
@@ -469,9 +531,15 @@ public static boolean isMatch(AttributeDescriptor a, AttributeDescriptor b) {
             return true;
         }
 
-        if (a.getLocalName().equals(b.getLocalName())
-                && a.getType().getBinding().equals(b.getType().getBinding())) {
-            return true;
+        if (strict) {
+            if (a.getLocalName().equals(b.getLocalName())
+                    && a.getType().getBinding().equals(b.getType().getBinding())) {
+                return true;
+            }
+        } else {
+            if (a.getLocalName().equals(b.getLocalName()) && a.getClass().equals(b.getClass())) {
```

Comment:
I'm not sure why we want to compare the descriptor classes, outside of maybe backwards compatibility with what looked like a simple coding mistake (to me)... AttributeDescriptor is an interface, two different implementations for it should effectively behave the same way. If what you're looking for is a comparison that only accounts for name, I'd limit it to it.

Reply:
Ah, now I see the test... so the comparison would also fail between a common attribute vs a geometry attribute, AttributeDescriptor vs GeometryDescriptor. Uh... still wondering if this is useful though, do you have a specific use case where this happens? I guess you must have one that prompted this pull request to start with :-D

Reply:
The motivation behind the present PR should by sought to the enhancement introduced recently in udig via https://github.com/locationtech/udig-platform/pull/218 and the ability provided by eclipse framework to define dynamic adapters that adapt/transform between java types essentially without limitations.
The changes introduced by GEOT-4931 and its backporting will render this feature useless and we need to fully rewrite it while by adding the extra proposed comparison method will require only minor modifications (only shifting to use the new compareNames method)

Reply:
Ok, and is the comparison of descriptor class necessary for that (sorry, I have no time to read into the udig ticket)? Method says "compareNames" but in reality it's comparing names and descriptor classes

Reply:
Provides for an extra level of 'logical' safety distinguishing between geometry filelds and standard attribute fields. If we have a GeometryDescriptor with name 'name1' in a shapefile and an AttributeDescriptor with the same name in another file this featuresTypes should definately not match and no adaptation is possible.
If method naming is the problem then propose a more convenient name and I will accept it :)

Reply:
Right but it should not distinguish between two impelmentations of "normal attributes" imho... if you don't really need it, I'd suggest to remove it and be true to a simple method name "compareNames", as opposed to a "compareNamesAndDescriptorClass". If it's hard to explain, unless there is a very good reason to go there, avoid it.

Reply:
I would prefer to be backward compatible and in general I do not like not to be able to distingush between an AttributeDescriptor,  GeometryDescriptor etc but you may be right in the sense that since the name of the method is compareNames() we must do just that.
Therefore, I would modify line 540 and correct appropriately the javadoc comments.